OutOfSchool/OutOfSchool.WebApi/Services/Database/WorkshopService.cs

```diff
@@ -266,22 +269,100 @@ public async Task<SearchResult<T>> GetByProviderId<T>(Guid id, ExcludeIdFilter f
             filter.ExcludedId == null
                 ? (x.ProviderId == id)
                 : (x.ProviderId == id && x.Id != filter.ExcludedId)).ConfigureAwait(false);
-        var workshops = await workshopRepository.Get(
+
+        var workshops = workshopRepository.Get(
                 skip: filter.From,
                 take: filter.Size,
                 includeProperties: includingPropertiesForMappingDtoModel,
                 where: x => filter.ExcludedId == null
                     ? (x.ProviderId == id)
-                    : (x.ProviderId == id && x.Id != filter.ExcludedId))
-            .ToListAsync()
-            .ConfigureAwait(false);
-
-        logger.LogInformation(!workshops.Any()
+                    : (x.ProviderId == id && x.Id != filter.ExcludedId));
+
+        var chatrooms = roomRepository.Get(
+            includeProperties: $"{nameof(ChatRoomWorkshop.ChatMessages)}",
+            where: x => x.ChatMessages.Where(x => x.ReadDateTime == null && !x.SenderRoleIsProvider).Any());
+
+        var query = from w in workshops
+                    join c in chatrooms
+                    on w.Id equals c.WorkshopId
+                    into g
+                    from cg in g.DefaultIfEmpty()
+                    select new
+                    {
+                        w.Id,
+                        w.Title,
+                        w.Phone,
+                        w.Email,
+                        w.Website,
+                        w.Facebook,
+                        w.Instagram,
+                        w.MinAge,
+                        w.MaxAge,
+                        w.CompetitiveSelection,
+                        w.CompetitiveSelectionDescription,
+                        w.Price,
+                        w.WorkshopDescriptionItems,
+                        w.WithDisabilityOptions,
+                        w.DisabilityOptionsDesc,
+                        w.CoverImageId,
+                        w.ProviderTitle,
+                        w.ProviderOwnership,
+                        w.Keywords,
+                        w.PayRate,
+                        w.ProviderId,
+                        w.AddressId,
+                        w.InstitutionHierarchyId,
+                        w.Status,
+                        w.AvailableSeats,
+                        w.Provider,
+                        w.InstitutionHierarchy,
+                        w.Address,
+                        w.ProviderAdmins,
+                        w.Teachers,
+                        w.Applications,
+                        w.DateTimeRanges,
+                        ChatRooms = cg,
```

Comment:
(Performance, readability, ...) The query is "irrelevant". 

`ChatRooms` are never used. 

From all this grouping story you need only count! of cg.ChatMessages. Please check the query generated by [GroupBy](https://learn.microsoft.com/en-us/ef/core/querying/complex-query-operators#groupby) that is expected to be executed on the server side. If it is not doing it on server side (I'm bit suspicious about `chatrooms` filter) - find your LINQ way or finally write that piece of sql code.

Reply:
Thank you for your patience and detailed description of my bad😔 code. I tried to fix it.

Reply:
Np and thanks for fixes. Have you checked generated SQL?

Reply:
Yes, on Saturday I found it large with a duplicate subquery, but I thought it's ok🤔 (at 12 am). Today, after your question, I double-checked and understood where the duplicate subquery comes from - "whereExpression" in roomRepository.Get. 
I remove "whereExpression" and got SQL:
SELECT `c`.`WorkshopId`, COUNT(*) AS `UnreadMessageCount`
FROM `ChatRoomWorkshops` AS `c`
INNER JOIN (
    SELECT `c0`.`ChatRoomId`, `c0`.`ReadDateTime`, `c0`.`SenderRoleIsProvider`
    FROM `ChatMessageWorkshops` AS `c0`
    WHERE NOT (`c0`.`IsDeleted`)
) AS `t` ON `c`.`Id` = `t`.`ChatRoomId`
WHERE NOT (`c`.`IsDeleted`) AND (`t`.`ReadDateTime` IS NULL AND NOT (`t`.`SenderRoleIsProvider`))
GROUP BY `c`.`WorkshopId`
So, I will commit removing unnecessary whereExpression.

Reply:
"Midnight coding" ))
That's cool that you've found it! Hope it wasn't too boring

Reply:
I'm interested) Thanks for your mentorship!